lib/runtime-core/src/instance.rs

```diff
@@ -164,88 +164,107 @@ impl Instance {
     /// ```
     /// # use wasmer_runtime_core::{Func, Instance, error::ResolveResult};
     /// # fn typed_func(instance: Instance) -> ResolveResult<()> {
-    /// let func: Func<(i32, i32)> = instance.func("foo")?;
+    ///   let func: Func<(i32, i32)> = instance.func("foo")?;
+    ///   func.call(42, 43);
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn func<Args, Rets>(&self, name: &str) -> ResolveResult<Func<Args, Rets, Wasm>>
+    where
+        Args: WasmTypeList,
+        Rets: WasmTypeList,
+    {
+        let func_index = self.resolve_func(name)?;
+
+        self.func_with_index(func_index)
+    }
+
+    /// Through generic magic and the awe-inspiring power of traits, we bring you...
+    ///
+    /// # "Func"
+    ///
+    /// A [`Func`] allows you to call functions exported from wasm with
+    /// near zero overhead.
+    ///
+    /// [`Func`]: struct.Func.html
+    /// # Usage:
+    ///
+    /// ```
+    /// # use wasmer_runtime_core::{Func, Instance, error::ResolveResult};
+    /// # fn typed_func(instance: Instance) -> ResolveResult<()> {
+    ///
+    /// let func_index = instance.resolve_func("foo")?;
+    /// let func: Func<(i32, i32)> = instance.func_with_index(func_index)?;
     ///
     /// func.call(42, 43);
     /// # Ok(())
     /// # }
     /// ```
-    pub fn func<Args, Rets>(&self, name: &str) -> ResolveResult<Func<Args, Rets, Wasm>>
+    pub fn func_with_index<Args, Rets>(
+        &self,
+        func_index: usize,
+    ) -> ResolveResult<Func<Args, Rets, Wasm>>
     where
         Args: WasmTypeList,
         Rets: WasmTypeList,
     {
-        let export_index =
-            self.module
-                .info
-                .exports
-                .get(name)
-                .ok_or_else(|| ResolveError::ExportNotFound {
-                    name: name.to_string(),
-                })?;
+        let func_index = FuncIndex::new(func_index);
```

Comment:
I'm undecided on whether or not this function should take a `FuncIndex` or a `usize`.  The reason we use typed indices is so that numbers don't get mixed up.  Perhaps the context here is strong enough that we don't need the extra type safety. 

Hmmm, `FuncIndex::new` is at least undocumented and perhaps not public.  Then I suppose it's a matter of if this function should be called with arbitrary user-supplied numbers or if the `FuncIndex` must come from somewhere.  If the `FuncIndex` comes from a validated source then my comments about the `expect` regarding the `invariant` are incorrect because we'd be validating when handing out the `FuncIndex`s (though I'm not sure that's possible to do in general).

I'd like to hear some more opinions/arguments on this before shipping!